core/types/celo_legacy_tx.go

```diff
@@ -0,0 +1,162 @@
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package types
+
+import (
+	"io"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+var ethCompatibleTxNumFields = 9
+
+// ethCompatibleTxRlpList is used for RLP encoding/decoding of eth-compatible transactions.
+// As such, it:
+// (a) excludes the Celo-only fields,
+// (b) doesn't need the Hash or EthCompatible fields, and
+// (c) doesn't need the `json` or `gencodec` tags
+type ethCompatibleTxRlpList struct {
+	Nonce    uint64          // nonce of sender account
+	GasPrice *big.Int        // wei per gas
+	Gas      uint64          // gas limit
+	To       *common.Address `rlp:"nil"` // nil means contract creation
+	Value    *big.Int        // wei amount
+	Data     []byte          // contract invocation input data
+	V, R, S  *big.Int        // signature values
+}
+
+// celoTxRlpList is used for RLP encoding/decoding of celo transactions.
+type celoTxRlpList struct {
+	Nonce               uint64          // nonce of sender account
+	GasPrice            *big.Int        // wei per gas
+	Gas                 uint64          // gas limit
+	FeeCurrency         *common.Address `rlp:"nil"` // nil means native currency
+	GatewayFeeRecipient *common.Address `rlp:"nil"` // nil means no gateway fee is paid
+	GatewayFee          *big.Int        `rlp:"nil"`
+	To                  *common.Address `rlp:"nil"` // nil means contract creation
+	Value               *big.Int        // wei amount
+	Data                []byte          // contract invocation input data
+	V, R, S             *big.Int        // signature values
+}
+
+func toEthCompatibleRlpList(tx LegacyTx) ethCompatibleTxRlpList {
+	return ethCompatibleTxRlpList{
+		Nonce:    tx.Nonce,
+		GasPrice: tx.GasPrice,
+		Gas:      tx.Gas,
+		To:       tx.To,
+		Value:    tx.Value,
+		Data:     tx.Data,
+		V:        tx.V,
+		R:        tx.R,
+		S:        tx.S,
+	}
+}
+
+func toCeloRlpList(tx LegacyTx) celoTxRlpList {
+	return celoTxRlpList{
+		Nonce:    tx.Nonce,
+		GasPrice: tx.GasPrice,
+		Gas:      tx.Gas,
+		To:       tx.To,
+		Value:    tx.Value,
+		Data:     tx.Data,
+		V:        tx.V,
+		R:        tx.R,
+		S:        tx.S,
+
+		// Celo specific fields
+		FeeCurrency:         tx.FeeCurrency,
+		GatewayFeeRecipient: tx.GatewayFeeRecipient,
+		GatewayFee:          tx.GatewayFee,
+	}
+}
+
+func (tx *LegacyTx) setTxFromEthCompatibleRlpList(rlplist ethCompatibleTxRlpList) {
+	tx.Nonce = rlplist.Nonce
+	tx.GasPrice = rlplist.GasPrice
+	tx.Gas = rlplist.Gas
+	tx.To = rlplist.To
+	tx.Value = rlplist.Value
+	tx.Data = rlplist.Data
+	tx.V = rlplist.V
+	tx.R = rlplist.R
+	tx.S = rlplist.S
+	tx.Hash = nil // txdata.Hash is calculated and saved inside tx.Hash()
+
+	// Celo specific fields
+	tx.FeeCurrency = nil
+	tx.GatewayFeeRecipient = nil
+	tx.GatewayFee = big.NewInt(0)
+	tx.EthCompatible = true
+}
+
+func (tx *LegacyTx) setTxFromCeloRlpList(rlplist celoTxRlpList) {
```

Comment:
how about making it a standalone function, not to polute LegacyTx methods?

Reply:
Yep, seems a bit neater 👍